File: mergeSort.cc
// MergeSort.cpp : 定义控制台应用程序的入口点。
//

#include<iostream>
using namespace std;
#include "mergeSort.h"

int main() 
{
  MergeSort<int> *mergeSort=new MergeSort<int>(29);
  mergeSort->data[0]=4;  
  mergeSort->data[1]=26; 
  mergeSort->data[2]=22; 
  mergeSort->data[3]=13; 
  mergeSort->data[4]=5;  
  mergeSort->data[5]=28;  
  mergeSort->data[6]=6;  
  mergeSort->data[7]=27;  
  mergeSort->data[8]=9;  
  mergeSort->data[9]=10;
  mergeSort->data[10]=19;
  mergeSort->data[11]=12;
  mergeSort->data[12]=20;
  mergeSort->data[13]=16;
  mergeSort->data[14]=11;
  mergeSort->data[15]=18;
  mergeSort->data[16]=24;
  mergeSort->data[17]=1; 
  mergeSort->data[18]=21;
  mergeSort->data[19]=17;
  mergeSort->data[20]=2; 
  mergeSort->data[21]=14;
  mergeSort->data[22]=23; 
  mergeSort->data[23]=15;
  mergeSort->data[24]=0;
  mergeSort->data[25]=8;
  mergeSort->data[26]=7;
  mergeSort->data[27]=3;
  mergeSort->data[28]=25;
  mergeSort->Sort();
  mergeSort->Print();
  return 0;
}

File: src/MergeSort.java
class MergeSort {
	public int[] data;

	MergeSort(int[] data) {
		this.data = data;
	}

	void Initial() {

	}

	public void merge(int low, int mid, int high) {
		int h, i, j, k;
		int b[] = new int[50];
		h = low;
		i = low;
		j = mid + 1;

		while ((h <= mid) && (j <= high)) {
			if (data[h] <= data[j]) {
				b[i] = data[h];
				h++;
			} else {
				b[i] = data[j];
				j++;
			}
			i++;
		}
		if (h > mid) {
			for (k = j; k <= high; k++) {
				b[i] = data[k];
				i++;
			}
		} else {
			for (k = h; k <= mid; k++) {
				b[i] = data[k];
				i++;
			}
		}
		for (k = low; k <= high; k++)
			data[k] = b[k];
	}

	public void mergeEffecient(int low, int mid, int high) {
		int left, index, right;
		int newArr[] = new int[high - low + 1];
		left = low;
		index = 0;
		right = mid + 1;

		while (left <= mid && right <= high) {
			if (data[left] <= data[right]) {
				newArr[index++] = data[left++];
			} else {
				newArr[index++] = data[right++];
			}
		}
		if (left > mid) {
			for (int k = right; k <= high; k++) {
				newArr[index++] = data[k];
			}
		} else {
			for (int k = left; k <= mid; k++) {
				newArr[index++] = data[k];
			}
		}
		index = 0;
		for (int k = low; k <= high; k++)
			data[k] = newArr[index++];
	}

	public void mergeSort() {
		// TODO Auto-generated method stub
		mergeSort(0, data.length - 1);
	}

	void mergeSort(int left, int right) {
		if (left < right) {
			int pivot = (left + right) / 2;
			mergeSort(left, pivot);
			mergeSort(pivot + 1, right);
			// mergeEffecient(left, pivot, right);
			mergeEffecient(left, pivot, right);
			// System.out.print("Now mergeSort," + left + "," + right + "->");
			// Print(left, right);

		}
	}

	int Partition(int left, int right) {
		return (left + right) / 2;
	}

	void Print() {
		for (int i = 0; i < data.length; i++) {
			System.out.print(data[i] + ",");
		}
		System.out.println();
	}

	void Print(int l, int r) {
		for (int i = l; i <= r; i++) {
			System.out.print(data[i] + ",");
		}
		System.out.println();
	}
}
